addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/model/UniFiController.java

```diff
@@ -0,0 +1,300 @@
+    }
+
+    public void refresh() throws UniFiException {
+        synchronized (this) {
+            sitesCache = getSites();
+            devicesCache = getDevices();
+            clientsCache = getClients();
+            insightsCache = getInsights();
+        }
+    }
+
+    // Site API
+
+    public @Nullable UniFiSite getSite(@Nullable String id) {
+        UniFiSite site = null;
+        if (StringUtils.isNotBlank(id)) {
+            synchronized (this) {
+                site = sitesCache.get(id);
+            }
+            if (site == null) {
+                logger.debug("Could not find a matching site for id = '{}'", id);
+            }
+        }
+        return site;
+    }
+
+    // Device API
+
+    public @Nullable UniFiDevice getDevice(@Nullable String id) {
+        UniFiDevice device = null;
+        if (StringUtils.isNotBlank(id)) {
+            synchronized (this) {
+                device = devicesCache.get(id);
+            }
+            if (device == null) {
+                logger.debug("Could not find a matching device for id = '{}'", id);
+            }
+        }
+        return device;
+    }
+
+    // Client API
+
+    public @Nullable UniFiClient getClient(@Nullable String id) {
+        UniFiClient client = null;
+        if (StringUtils.isNotBlank(id)) {
+            synchronized (this) {
+                // mgb: first check active clients and fallback to insights if not found
+                client = clientsCache.get(id);
+                if (client == null) {
+                    client = insightsCache.get(id);
+                }
+            }
+            if (client == null) {
+                logger.debug("Could not find a matching client for id = {}", id);
+            }
+        }
+        return client;
+    }
+
+    protected void block(UniFiClient client, boolean blocked) throws UniFiException {
+        UniFiControllerRequest<Void> req = newRequest(Void.class);
+        req.setPath("/api/s/" + client.getSite().getName() + "/cmd/stamgr");
+        req.setBodyParameter("cmd", blocked ? "block-sta" : "unblock-sta");
+        req.setBodyParameter("mac", client.getMac());
+        executeRequest(req);
+    }
+
+    protected void reconnect(UniFiClient client) throws UniFiException {
+        UniFiControllerRequest<Void> req = newRequest(Void.class);
+        req.setPath("/api/s/" + client.getSite().getName() + "/cmd/stamgr");
+        req.setBodyParameter("cmd", "kick-sta");
+        req.setBodyParameter("mac", client.getMac());
+        executeRequest(req);
+    }
+
+    // Internal API
+
+    private <T> UniFiControllerRequest<T> newRequest(Class<T> responseType) {
+        return new UniFiControllerRequest<T>(responseType, gson, httpClient, host, port);
+    }
+
+    private <T> T executeRequest(UniFiControllerRequest<T> request) throws UniFiException {
+        T result;
+        try {
+            result = request.execute();
+        } catch (UniFiExpiredSessionException e) {
+            login();
```

Comment:
`login` calls `executeRequest` which means just looking at the code it can result in an endless loop. I don't know if it is practically possible, but it might be good to guard against an endless loop.

Reply:
The expired session exception should _theoretically_ only ever get thrown if you've already successfully logged in but your refresh value is so high that the session times out (will almost never happen).